File: core/scanner/tcp.go
package scanner

import (
	"context"
	"fmt"
	"net"
	"sync"
	"time"

	"github.com/netprowl/core/types"
)

// WHITE_PORTS 白名单端口（微信小程序允许）
var WHITE_PORTS = []int{80, 443, 8080, 8443, 554, 5000, 9000, 49152}

// TCPConfig TCP 扫描配置
type TCPConfig struct {
	Ports       []int // 目标端口，为空则使用 WHITE_PORTS
	Concurrency int   // 并发连接数，默认 100
	TimeoutMs   int   // 单端口超时(ms)，默认 2000
}

// DefaultTCPConfig 默认配置（白名单端口）
var DefaultTCPConfig = TCPConfig{
	Ports:       WHITE_PORTS,
	Concurrency: 100,
	TimeoutMs:   2000,
}

const (
	defaultTimeout = 2 * time.Second
	maxConcurrency = 200
)

// PortProbe is the scanning result for a single port probe.
// Renamed from ScanResult to avoid confusion with types.ScanResult.
type PortProbe struct {
	IP    string
	Port  int
	State string // "open" | "closed"
	Banner string
}

// ProbeTCPPorts 扫描单个 IP 的端口
func ProbeTCPPorts(ctx context.Context, ip string, cfg TCPConfig) ([]types.Port, error) {
	if len(cfg.Ports) == 0 {
		cfg.Ports = WHITE_PORTS
	}

	timeout := time.Duration(cfg.TimeoutMs) * time.Millisecond
	if timeout == 0 {
		timeout = defaultTimeout
	}

	concurrency := cfg.Concurrency
	if concurrency <= 0 {
		concurrency = 100
	}

	results := make([]PortProbe, 0)
	var mu sync.Mutex
	var wg sync.WaitGroup
	sem := make(chan struct{}, concurrency)

	for _, port := range cfg.Ports {
		select {
		case <-ctx.Done():
			return nil, ctx.Err()
		case sem <- struct{}{}:
		}

		wg.Add(1)
		go func(port int) {
			defer wg.Done()
			defer func() { <-sem }()

			result := probePort(ip, port, timeout)
			mu.Lock()
			if result.State == "open" {
				results = append(results, result)
			}
			mu.Unlock()
		}(port)
	}
	wg.Wait()

	return buildPortList(results), nil
}

// ProbeTCPPort 探测单个端口
func ProbeTCPPort(ctx context.Context, ip string, port int, timeout time.Duration) (types.Port, error) {
	result := probePort(ip, port, timeout)
	if result.State != "open" {
		return types.Port{Port: port, State: types.PortStateClosed}, nil
	}
	return types.Port{
		Port:    port,
		State:   types.PortStateOpen,
		Service: GuessService(port),
		Banner:  result.Banner,
	}, nil
}

func probePort(ip string, port int, timeout time.Duration) PortProbe {
	result := PortProbe{IP: ip, Port: port, State: "closed"}
	conn, err := net.DialTimeout("tcp", fmt.Sprintf("%s:%d", ip, port), timeout)
	if err != nil {
		return result
	}
	defer conn.Close()
	result.State = "open"
	result.Banner = grabBanner(conn, port)
	return result
}

func grabBanner(conn net.Conn, port int) string {
	conn.SetDeadline(time.Now().Add(1 * time.Second))
	switch port {
	case 80, 8080, 8443:
		fmt.Fprint(conn, "HEAD / HTTP/1.0\r\n\r\n")
	}
	buf := make([]byte, 1024)
	n, _ := conn.Read(buf)
	if n > 0 {
		return string(buf[:n])
	}
	return ""
}

func buildPortList(results []PortProbe) []types.Port {
	ports := make([]types.Port, 0, len(results))
	for _, r := range results {
		ports = append(ports, types.Port{
			Port:    r.Port,
			State:   types.PortState(r.State),
			Service: GuessService(r.Port),
			Banner:  r.Banner,
		})
	}
	return ports
}

var serviceMap = map[int]string{
	80:   "http",
	443:  "https",
	22:   "ssh",
	21:   "ftp",
	25:   "smtp",
	110:  "pop3",
	143:  "imap",
	135:  "msrpc",
	139:  "netbios",
	445:  "smb",
	3389: "rdp",
	8080: "http-alt",
	8443: "https-alt",
	5000: "upnp",
	9000: "cslistener",
	554:  "rtsp",
}

func GuessService(port int) string {
	if s, ok := serviceMap[port]; ok {
		return s
	}
	return "unknown"
}

File: core/src/scanner/tcp.rs
use std::collections::HashMap;
use std::time::Duration;
use std::sync::Arc;

use tokio::io::{AsyncReadExt, AsyncWriteExt};
use tokio::net::TcpStream;
use tokio::sync::Semaphore;

use crate::types::{Device, Port, PortState};

lazy_static::lazy_static! {
    static ref SERVICE_MAP: HashMap<u16, &'static str> = {
        let mut m = HashMap::new();
        m.insert(21, "ftp");
        m.insert(22, "ssh");
        m.insert(23, "telnet");
        m.insert(25, "smtp");
        m.insert(53, "dns");
        m.insert(80, "http");
        m.insert(110, "pop3");
        m.insert(143, "imap");
        m.insert(443, "https");
        m.insert(554, "rtsp");
        m.insert(5000, "upnp");
        m.insert(8080, "http-proxy");
        m.insert(8443, "https-alt");
        m.insert(3306, "mysql");
        m.insert(5432, "postgresql");
        m.insert(6379, "redis");
        m.insert(27017, "mongodb");
        m.insert(9200, "elasticsearch");
        m
    };
}

/// Scan a single TCP port with timeout
async fn scan_port(
    ip: &str,
    port: u16,
    timeout: Duration,
) -> Port {
    let addr = format!("{}:{}", ip, port);
    let state = match tokio::time::timeout(timeout, TcpStream::connect(&addr)).await {
        Ok(Ok(_)) => PortState::Open,
        Ok(Err(_)) => PortState::Closed,
        Err(_) => PortState::Filtered,
    };

    let service = SERVICE_MAP.get(&port).map(|s: &&'static str| s.to_string());

    Port {
        port,
        state,
        service,
        banner: None,
    }
}

/// Scan multiple TCP ports with concurrency control
pub async fn scan_tcp(
    ip: &str,
    ports: &[u16],
    timeout_ms: u64,
    concurrency: u32,
) -> Vec<Port> {
    let timeout = Duration::from_millis(timeout_ms);
    let semaphore = Arc::new(Semaphore::new(concurrency as usize));
    let mut handles = Vec::new();

    for &port in ports {
        let permit = semaphore.clone().acquire_owned().await.unwrap();
        let ip_owned = ip.to_string();

        let handle = tokio::spawn(async move {
            let port_result = scan_port(&ip_owned, port, timeout).await;
            drop(permit);
            port_result
        });

        handles.push(handle);
    }

    let mut results = Vec::new();
    for handle in handles {
        if let Ok(port) = handle.await {
            results.push(port);
        }
    }

    results.sort_by_key(|p| p.port);
    results
}

/// Scan a subnet for open ports
pub async fn scan_subnet(
    subnet: &str,
    ports: Vec<u16>,
    timeout_ms: u64,
    concurrency: u32,
) -> Vec<Device> {
    let base = if subnet.ends_with(".0") || subnet.ends_with(".0/24") {
        &subnet[..subnet.len() - 2]
    } else {
        subnet
    };

    let mut devices = Vec::new();

    for i in 1..=254 {
        let ip = format!("{}.{}", base, i);
        let ip_clone = ip.clone();
        let ports_clone = ports.clone();

        let handle = tokio::spawn(async move {
            let open_ports = scan_tcp(&ip_clone, &ports_clone, timeout_ms, concurrency).await;
            let open_ports: Vec<_> = open_ports.into_iter().filter(|p| p.state == PortState::Open).collect();

            if !open_ports.is_empty() {
                let mut device = Device::new(ip_clone);
                device.open_ports = open_ports;
                Some(device)
            } else {
                None
            }
        });

        if let Ok(Some(device)) = handle.await {
            devices.push(device);
        }
    }

    devices
}

/// Get service name for a port
pub fn get_service_name(port: u16) -> Option<String> {
    SERVICE_MAP.get(&port).map(|s: &&'static str| s.to_string())
}

/// Grab banner from an open port
pub async fn grab_banner(
    ip: &str,
    port: u16,
    timeout: Duration,
) -> Option<String> {
    let addr = format!("{}:{}", ip, port);

    let mut stream = match tokio::time::timeout(timeout, TcpStream::connect(&addr)).await {
        Ok(Ok(s)) => s,
        _ => return None,
    };

    // Send HTTP HEAD for web ports
    if matches!(port, 80 | 8080 | 8443 | 5000 | 554) {
        let _ = stream
            .write_all(b"HEAD / HTTP/1.0\r\nHost: localhost\r\n\r\n")
            .await;
    }

    let mut buf = [0u8; 1024];
    let n = match tokio::time::timeout(timeout, stream.read(&mut buf)).await {
        Ok(Ok(n)) => n,
        _ => return None,
    };

    if n > 0 {
        let banner = String::from_utf8_lossy(&buf[..n]).trim().to_string();
        if !banner.is_empty() {
            return Some(banner);
        }
    }

    None
}
